### PR TITLE
Cleanup and tests for "merge left" behaviour

### PR DESCRIPTION
**Pull request by jml**
I had an idle hour, so I extracted the `mergeLeftBy` behaviour out of the `annotateTokens` function in an effort to clarify what we're doing. "No raw algorithms" and all that.

* New `mergeLeftBy` function
* Some types now derive `Show`, so we can report errors
* QuickCheck and unit tests for `mergeLeftBy`
* `annotateTokens` uses `mergeLeftBy` (still untested)
* New `assertRight` helper

[Review on Reviewable](https://reviewable.io/reviews/jml/holborn/9)

### CODE REVIEW

File: holborn-web/lib/Holborn/Web.hs [hunk 1]
```diff
@@ -27,16 +28,26 @@ import Holborn.Types (Annotation(..), Symbol(..), HolbornToken(..), Reference(..
...
 
+
 annotateTokens :: [Token] -> Annotation -> [HolbornToken]
--- left merge
-annotateTokens [] (Annotation []) = []
-annotateTokens (x:xs) (Annotation (y@(Symbol symbolName symbolRef):ys))
-  | (tText x) == symbolName = (HolbornToken x symbolRef) : (annotateTokens xs (Annotation ys))
-  | otherwise = (HolbornToken x (Reference "")) : annotateTokens xs (Annotation (y:ys))
+annotateTokens tokens (Annotation annotation) =
+  map (uncurry mergeTokens) (assertRight "Could not match AST data to tokenized data" mergeResult)
```
Comment (teh) [inline holborn-web/lib/Holborn/Web.hs]: not sure I like uncurry over just writing the thing out but I guess that's the haskell way of life :)

**Comment by teh** [comment]
nice, LGTM

I think the algorithm is in the same O-class but could be slower when written as a left-merge with Nothings + filter like right now but that doesn't really matter at all. Cleaner to read.

File: holborn-web/lib/Holborn/Web.hs (hunk 1 above)
Comment (jml) [inline holborn-web/lib/Holborn/Web.hs]: I'm not 100% sure either, fwiw. Using `uncurry` here makes the duplication in `mergeTokens` more obvious to me, and saves some parentheses.

**Comment by jml** [comment]
I think this is a map, not a filter. I'd actually be curious as to whether separating the match & merge steps makes things slower.

**Comment by teh** [comment]
Me too but more interested in progress for now :) Merge when ready!

**Pull request merged**